### PR TITLE
Generate: PT's `top_p` enforces `min_tokens_to_keep` when it is `1`

### LINKED ISSUE #23688: LlamaForCausalLM generate() runtime error when top_p=0
### System Info

- `transformers` version: 4.29.2
- Platform: Linux-4.4.0-210-generic-x86_64-with-debian-stretch-sid
- Python version: 3.7.3
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 1.13.1+cu117 (True)
- Tensorflow version (GPU?): 2.5.0 (True)

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

This example follows and modifies the LLaMA documentation at: https://huggingface.co/docs/transformers/v4.29.1/model_doc/llama.

```python
from transformers import AutoTokenizer, LlamaForCausalLM

model = LlamaForCausalLM.from_pretrained(PATH_TO_CONVERTED_WEIGHTS).cuda()
tokenizer = AutoTokenizer.from_pretrained(PATH_TO_CONVERTED_TOKENIZER)

prompt = "Hey, are you conscious? Can you talk to me?"
inputs = tokenizer(prompt, return_tensors="pt").cuda()

# Generate
generate_ids = model.generate(
    inputs.input_ids, max_length=1024,
    do_sample=True,
    temperature=0.6,
    top_k=1000,
    top_p=0.0,  # cause error
    repetition_penalty=(1.0 / 0.85),
    )
tokenizer.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]
```

>RuntimeError: probability tensor contains either `inf`, `nan` or element < 0

### Expected behavior

`generate()` should check if `top_p==0` and disables it to avoid numerical error. This behavior is the same as when `top_p==1` and consistent with the documentation.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #23688 

Contrary to our description in the docstring, PT's `top_p` was not enforcing `min_tokens_to_keep` when it was 1 (the default). TF and FLAX were fine. This PR corrects it, and adds a check on `min_tokens_to_keep` (must be a non-negative integer)

### CODE REVIEW

**Comment by amyeroberts** [comment]
> top_p was not enforcing min_tokens_to_keep when it was 1 

From the diff - I don't see how this is resolved. The checks ensure the value of `min_tokens_to_keeps` but doesn't seem to be conditional on `top_p`. Am I missing something?

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -266,9 +268,8 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
         # Remove tokens with cumulative top_p above the threshold (token with 0 are kept)
         sorted_indices_to_remove = cumulative_probs <= (1 - self.top_p)
-        if self.min_tokens_to_keep > 1:
```
Comment (gante) [inline src/transformers/generation/logits_process.py:269]: @amyeroberts -- this line was preventing the application of `min_tokens_to_keep` when it was `1`. Removing it sorts the problem :)

Comment (gante) [reply src/transformers/generation/logits_process.py:269]: e.g. if we were to set `top_p=0.0`, `.generate()` would crash due to the lack of suitable continuations, despite the default of `min_tokens_to_keep` being `1`. 

After this fix, setting `top_p=0.0` does not crash the code. Refresher: Top P selects all the tokens whose cumulative probability is `>= top_p`. Setting it to `0.0` means it should pick exactly one token.

Comment (amyeroberts) [reply src/transformers/generation/logits_process.py:269]: Ah, it's a stupid ambiguity of English issue :D From the title I thought it meant 'when `top_p` is 1`.

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

**Comment by njhill** [comment]
@gante I hit an issue related to this in the prior version of transformers, glad to see that it's fixed thanks! However why don't we enforce `min_tokens_to_keep >= 1`. 0 makes no sense right?

**Comment by gante** [comment]
@njhill true, the initial check should be against `>=1`, patching it

**Pull request merged**